File: solutions/go/0017.go
package main

import (
    "strings"
)

func letterCombinations(digits string) []string {
    if digits == "" {
        return []string{}
    }

    digitMap := map[byte]string{
        '2': "abc",
        '3': "def",
        '4': "ghi",
        '5': "jkl",
        '6': "mno",
        '7': "pqrs",
        '8': "tuv",
        '9': "wxyz",
    }

    var result []string
    var combination []string

    var backtrack func(index int)
    backtrack = func(index int) {
        if index == len(digits) {
            result = append(result, strings.Join(combination, ""))
            return
        }

        letters := digitMap[digits[index]]
        for i := 0; i < len(letters); i++ {
            combination = append(combination, string(letters[i]))
            backtrack(index + 1)
            combination = combination[:len(combination)-1]
        }
    }

    backtrack(0)
    return result
}

File: solutions/java/0017.java
class Solution {
    public List<String> letterCombinations(String digits) {
        if (digits.isEmpty()) {
            return new ArrayList<>();
        }

        Map<Character, String> map = new HashMap<>();
        map.put('2', "abc");
        map.put('3', "def");
        map.put('4', "ghi");
        map.put('5', "jkl");
        map.put('6', "mno");
        map.put('7', "pqrs");
        map.put('8', "tuv");
        map.put('9', "wxyz");

        List<String> result = new ArrayList<>();
        StringBuilder combination = new StringBuilder();

        backtrack(digits, 0, map, combination, result);
        return result;
    }

    private void backtrack(String digits, int index, Map<Character, String> map, StringBuilder combination, List<String> result) {
        if (index == digits.length()) {
            result.add(combination.toString());
            return;
        }

        String letters = map.get(digits.charAt(index));
        for (char letter : letters.toCharArray()) {
            combination.append(letter);
            backtrack(digits, index + 1, map, combination, result);
            combination.deleteCharAt(combination.length() - 1);
        }
    }
}